src/rust/engine/process_execution/src/remote.rs

```diff
@@ -134,19 +134,33 @@ impl CommandRunner {
                                 )
                               )
                             })
-                            .map(|operation| future::Loop::Continue(operation))
+                            // Reset `iter_num` on `MissingDigests`
+                            .map(|operation| future::Loop::Continue((operation, 0)))
                             .to_boxed()
                       },
                       ExecutionError::NotFinished(operation_name) => {
                         let mut operation_request =
                           bazel_protos::operations::GetOperationRequest::new();
                         operation_request.set_name(operation_name);
+
+                        let max_wait = 5000;
+                        let backoff_period = min(max_wait, ((1 + iter_num) * 500));
+
                         let grpc_result = map_grpc_result(
                           operations_client.get_operation(&operation_request)
                         );
-                        future::ok(
-                          future::Loop::Continue(
-                            try_future!(grpc_result))).to_boxed() as BoxFuture<_, _>
+                        Delay::new(Duration::from_millis(backoff_period))
+                            .then(move |res| {
```

Comment:
Rather than using an unconditional `.then`, with a `match` in it:
```
.then(|res| match res {
  Ok(_) => ...,
  Err(e) => ...,
)
```

it may be a little cleaner to use `map`/`map_err`. Because our success case may fail (that's what the `try_future!` is doing - returning a failure case if `grpc_result` is an error), we need to use `and_then` rather than `map`:

```
Delay::new(...)
  .map_err(|err| err.to_string())
  .and_then(|_| future::ok(future::Loop::Continue((try_future!(grpc_result), iter_num + 1))).to_boxed() as BoxFuture<_, _>)
```

Then, noting that `Future::and_then` accepts something which returns an `IntoFuture`, we can actually just return a `Result`, and not need to worry about all of the `Future`-specific nonsense of `try_future!`, `to_boxed()` and so on:

```
Delay::new(...)
  .map_err(|err| err.to_string())
  .and_then(|_| Ok(future::Loop::Continue((grpc_result?, iter_num + 1))))
```

Reply:
1. Used the new pattern.

2. 
```
  .and_then(|_| Ok(future::Loop::Continue((grpc_result?, iter_num + 1))))
```
doesn't seem to work, so left `try_future!` untouched.